Add some named tensor helper functions

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21648 Expose torch.empty(sizes, *, names, ...) to Python
* #21647 Implement at::empty(IntArrayRef, DimnameList?, TensorOptions) in aten
* **#21636 Add some named tensor helper functions**
* #21632 Add some more namedtensor builds to the CI
* #21633 Add virtual dtor to NamedTensorMetaInterface

at::internal_set_names_inplace(Tensor, optional<DimnameList>)
- Used for setting the names of a tensor. This isn't a native function for
  performance reasons; it'll be used to implement tensor.set_names_,
  torch.empty(..., names=), and other functions.

optional<DimnameList> Tensor::names()
- Access names of a tensor.

Test Plan:
- New tests in [namedtensor ci]

gh-metadata: pytorch pytorch 21636 gh/zou3519/46/head

Differential Revision: [D15780269](https://our.internmc.facebook.com/intern/diff/D15780269)

zou3519 committed:
Add some named tensor helper functions

at::internal_set_names_inplace(Tensor, optional<DimnameList>)
- Used for setting the names of a tensor. This isn't a native function for
  performance reasons; it'll be used to implement tensor.set_names_,
  torch.empty(..., names=), and other functions.

optional<DimnameList> Tensor::names()
- Access names of a tensor.

Test Plan:
- New tests in [namedtensor ci]

gchanan commented:
@nairbv can you take a look?

nairbv commented on aten/src/ATen/templates/Tensor.h:
```diff
@@ -257,6 +267,7 @@ class CAFFE2_API Tensor {
 
   /// Returns a `Tensor`'s dimension names data structure
   NamedTensorMeta* get_named_tensor_meta() const;
+  NamedTensorMeta* get_named_tensor_meta();
```
c++ question .... why don't these function signatures conflict?

zou3519 replied:
c++ lets you overload member methods based on const-ness.

zou3519 replied:
Err I'm changing the signatures to `const NamedTensorMeta* get_named_tensor_meta() const;`. I am not sure if this is actually different.

nairbv commented on aten/src/ATen/test/NamedTensor_test.cpp:
```diff
@@ -63,4 +63,30 @@ TEST(NamedTensorTest, attachMetadata) {
+  auto N = Dimname::fromSymbol(Symbol::dimname("N"));
+  auto C = Dimname::fromSymbol(Symbol::dimname("C"));
+  auto H = Dimname::fromSymbol(Symbol::dimname("H"));
+  auto W = Dimname::fromSymbol(Symbol::dimname("W"));
```
should there be a Dimname::fromString just to save typing?

zou3519 replied:
Maybe. There's a side-effect every time one constructs a Symbol via `Symbol::dimname` (it gets added to the global c++ InternedStrings table) and I didn't want any methods on `Dimname` to have side effects.

zou3519 replied:
I'm going to add a helper function to make the typing less in this testing file

nairbv commented on aten/src/ATen/NamedTensor.cpp:
```diff
@@ -1,14 +1,37 @@
+  auto* meta = tensor.get_named_tensor_meta();
+  if (meta == nullptr) {
+    tensor.unsafeGetTensorImpl()->set_named_tensor_meta(
+        torch::make_unique<NamedTensorMeta>(*names));
```
another c++ question... is this going to copy the vector of names into a new vector? does the pass by value also copy (copying the list twice?). 

Would it be any better to pass the tensor by reference?

zou3519 replied:
`DimnameList` is defined to be `ArrayRef<Dimname>`. You can think of it like a `reference_wrapper<vector<Dimname>>` (so yes, it's already being passed by reference).

nairbv replied:
ah OK. And the Tensor argument?

zou3519 replied:
It's probably slightly faster to pass `Tensor` as `Tensor&`. `Tensor` itself is an `intrusive_ptr<TensorImpl>` I think; passing it as `Tensor` means we bump the refcount and copy the pointer (which is fast, because it's just a word) and passing it as `Tensor&` means there is no refcount bump.

nairbv replied:
I ask partly because it seems like we usually use `Tensor &`:
```
(base) [bvaughan@devgpu005.ash6 ~/repos/pytorch] gg -rnE '\(Tensor & [a-zA-Z]+.*\).*;' .  | grep \\.h   | wc -l
3941
(base) [bvaughan@devgpu005.ash6 ~/repos/pytorch] gg -rnE '\(Tensor [a-zA-Z]+.*\).*;' .  | grep \\.h   | wc -l
51
```

zou3519 replied:
Yup, I'll change it to `const Tensor&` (I think that is what we use for functions that modify tensors)

nairbv commented on aten/src/ATen/NamedTensor.cpp:
```diff
@@ -1,14 +1,37 @@
 bool NamedTensorMeta::has_names() const {
   return !std::all_of(
-      names.begin(), names.end(), [](const Dimname& n) {
+      names_.begin(), names_.end(), [](const Dimname& n) {
```
Are there cases where this would be called many times repeatedly on the same tensor without the names having changed? Any thoughts on if it could cause performance issues or if we would need to cache the result?

zou3519 replied:
Oh we could cache this. I don't want to write in too many optimizations right now in fear that they might be premature, so I'll leave that for the future.

zou3519 committed:
Update on "Add some named tensor helper functions"

Add some named tensor helper functions

at::internal_set_names_inplace(Tensor, optional<DimnameList>)
- Used for setting the names of a tensor. This isn't a native function for
  performance reasons; it'll be used to implement tensor.set_names_,
  torch.empty(..., names=), and other functions.

optional<DimnameList> Tensor::names()
- Access names of a tensor.

Test Plan:
- New tests in [namedtensor ci]

gh-metadata: pytorch pytorch 21636 gh/zou3519/46/head

nairbv reviewed (changes requested):
`Tensor &`

zou3519 committed:
Update on "Add some named tensor helper functions"

Add some named tensor helper functions

at::internal_set_names_inplace(Tensor, optional<DimnameList>)
- Used for setting the names of a tensor. This isn't a native function for
  performance reasons; it'll be used to implement tensor.set_names_,
  torch.empty(..., names=), and other functions.

optional<DimnameList> Tensor::names()
- Access names of a tensor.

Test Plan:
- New tests in [namedtensor ci]

gh-metadata: pytorch pytorch 21636 gh/zou3519/46/head

zou3519 commented:
@nairbv updated
